nodebuilder/header/data_commitment.go

```diff
@@ -0,0 +1,242 @@
+// For more information, refer to:
+// https://github.com/celestiaorg/blobstream-contracts/blob/master/src/DataRootTuple.sol
+func EncodeDataRootTuple(height uint64, dataRoot [32]byte) ([]byte, error) {
+	paddedHeight, err := To32PaddedHexBytes(height)
+	if err != nil {
+		return nil, err
+	}
+	return append(paddedHeight, dataRoot[:]...), nil
+}
+
+// dataCommitmentBlocksLimit The maximum number of blocks to be used to create a data commitment.
+// It's a local parameter to protect the API from creating unnecessarily large commitments.
+const dataCommitmentBlocksLimit = 10_000 // ~33 hours of blocks assuming 12-second blocks.
+
+// validateDataCommitmentRange runs basic checks on the asc sorted list of
+// heights that will be used subsequently in generating data commitments over
+// the defined set of heights.
+func (s *Service) validateDataCommitmentRange(ctx context.Context, start, end uint64) error {
+	if start == 0 {
+		return fmt.Errorf("the start block is 0")
+	}
+	if start >= end {
+		return fmt.Errorf("end block is smaller or equal to the start block")
+	}
+	heightsRange := end - start
+	if heightsRange > uint64(dataCommitmentBlocksLimit) {
+		return fmt.Errorf("the query exceeds the limit of allowed blocks %d", dataCommitmentBlocksLimit)
+	}
+
+	currentHeader, err := s.NetworkHead(ctx)
+	if err != nil {
+		return err
+	}
+	// the data commitment range is end exclusive
+	if end > currentHeader.Height()+1 {
+		return fmt.Errorf(
+			"end block %d is higher than current chain height %d",
+			end,
+			currentHeader.Height(),
+		)
+	}
+
+	currentLocalHeader, err := s.LocalHead(ctx)
+	if err != nil {
+		return err
+	}
+	// the data commitment range is end exclusive
+	if end > currentLocalHeader.Height()+1 {
+		return fmt.Errorf(
+			"end block %d is higher than local chain height %d. Wait for the node until it syncs up to %d",
+			end,
+			currentLocalHeader.Height(),
+			end,
+		)
+	}
+	return nil
+}
+
+// hashDataRootTuples hashes a list of blocks data root tuples, i.e., height, data root and square
+// size, then returns their merkle root.
+func hashDataRootTuples(tuples []DataRootTuple) ([]byte, error) {
+	if len(tuples) == 0 {
+		return nil, fmt.Errorf("cannot hash an empty list of data root tuples")
+	}
+	dataRootEncodedTuples := make([][]byte, 0, len(tuples))
+	for _, tuple := range tuples {
+		encodedTuple, err := EncodeDataRootTuple(
+			tuple.height,
+			tuple.dataRoot,
+		)
+		if err != nil {
+			return nil, err
+		}
+		dataRootEncodedTuples = append(dataRootEncodedTuples, encodedTuple)
+	}
+	root := merkle.HashFromByteSlices(dataRootEncodedTuples)
+	return root, nil
+}
+
+// validateDataRootInclusionProofRequest validates the request to generate a data root
+// inclusion proof.
+func (s *Service) validateDataRootInclusionProofRequest(
+	ctx context.Context,
+	height, start, end uint64,
+) error {
+	err := s.validateDataCommitmentRange(ctx, start, end)
```

Comment:
can you combine these two methods please, it's fine if it's a bit longer of a function.

Reply:
the validation is used twice. If we remove that function and put the logic here, we will have to maintain two validation logic. I think it's better in a separate method